// @prismicio/mocks/lib/generators/widgets/nestable/Link/MediaLink.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaLinkMock = void 0;
const content_1 = require("@prismicio/types-internal/lib/content");
const utils_1 = require("../../../utils");
exports.MediaLinkMock = {
    generate(def, config) {
        var _a, _b, _c, _d;
        let value;
        if (config === null || config === void 0 ? void 0 : config.value) {
            value = config.value;
        }
        else {
            const text = ((_a = def.config) === null || _a === void 0 ? void 0 : _a.allowText) ? (0, utils_1.slug)() : undefined;
            const variant = (_c = (_b = def.config) === null || _b === void 0 ? void 0 : _b.variants) === null || _c === void 0 ? void 0 : _c[Math.floor(Math.random() * def.config.variants.length)];
            value = (_d = config === null || config === void 0 ? void 0 : config.value) !== null && _d !== void 0 ? _d : {
                __TYPE__: content_1.FileLinkType,
                id: "mock_Media_link",
                kind: "file",
                name: "mock_name",
                url: "https://source.unsplash.com/daily",
                size: "0",
                ...(text && { text }),
                ...(variant && { variant }),
            };
        }
        return content_1.FileLinkContent.encode(value);
    },
};
