// @prismicio/api-renderer/lib/widgets/nestable/Link/ExternalLinkRenderer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const content_1 = require("@prismicio/types-internal/lib/content");
const ExternalLinkRenderer = {
    renderV1(link, _fetch, options) {
        return {
            type: "Link.web",
            value: {
                ...(!(options === null || options === void 0 ? void 0 : options.omitKey) && { key: link.key }),
                ...(content_1.FilledExternalLinkContent.is(link.value) && {
                    url: link.value.url,
                    ...(typeof link.value.target === "string" && { target: link.value.target }),
                }),
                ...(link.value.text && { text: link.value.text }),
                ...(link.value.variant && { variant: link.value.variant }),
            },
        };
    },
    renderV2(def, link, _fetch, options) {
        const defVariants = def.variants || [];
        const linkVariant = link.value.variant && defVariants.includes(link.value.variant) ? link.value.variant : defVariants[0];
        return {
            link_type: "Web",
            ...(!(options === null || options === void 0 ? void 0 : options.omitKey) && { key: link.key }),
            ...(content_1.FilledExternalLinkContent.is(link.value) && {
                url: link.value.url,
                ...(typeof link.value.target === "string" && { target: link.value.target }),
            }),
            ...(def.allowText && link.value.text && { text: link.value.text }),
            ...(linkVariant && { variant: linkVariant }),
        };
    },
    renderMocks(def, link, options) {
        return this.renderV2(def, link, undefined, options);
    },
};
exports.default = ExternalLinkRenderer;
